packages/core-v2-graphql/src/models/Token.ts

```diff
@@ -1,4 +1,67 @@
+import { ERC20__factory } from "@elementfi/core-v2-typechain";
 import { CoreV2Context } from "src/context";
+import { TokenContract } from "src/datasources/TokenContract";
 import { Token } from "src/generated";
+import { getDataSourceByAddress } from "src/utils/getDataSourceByAddress";
 
-export const TokenModel = {};
+// model should be a class that takes in a datasource(s)
+export const TokenModel = {
+  getAllowance,
+  getBalanceOf,
+  getByAddress,
+};
+
+interface GetByAddressOptions {
+  address: string;
+  context: CoreV2Context;
+}
+
+async function getByAddress({
+  address,
+  context,
+}: GetByAddressOptions): Promise<Token> {
+  const tokenContract = getDataSourceByAddress(
+    address,
+    context.elementDataSources,
+    "tokenContracts",
+  );
+
+  return {
+    address: await tokenContract.address,
+    decimals: await tokenContract.decimals(),
+    symbol: await tokenContract.symbol(),
+    name: await tokenContract.name(),
+  };
+}
+
+async function getBalanceOf({
+  address,
+  owner,
+  context,
+}: { owner: string } & GetByAddressOptions): Promise<string> {
```

Comment:
The reusing of interfaces here might be a little more efficient, but also creates a pathway for spaghetti code in the future where updating one method's signature inadvertently affects multiple others.

Can we list each out and rather than taking the token address ( a field of the token) let's just be greedy and take the whole token. Then we can use `address` for the address we want to check the balance for. So our options would end up being:

```
{
  address: string,
  token: Token,
  context: CoreV2Contract
}
```

This creates a developer experience where I use a model to `get` an instance of something, then just pass around that instance:

```typescript
const fooToken = TokenModel.getByAddress({ address: "0x...", context: myContext })
const myFooTokenBalance = TokenModel.getAllowance({
  address: myAddress,
  token: fooToken,
  context: myContext
})
```

And I don't need to wonder which options are referring to fields of the token, I just pass the whole thing as `token`.

Reply:
Kind of confused on the first comment here. Are you suggesting I have a typed interface for each func sig  i.e. `getBalanceOptions`?

Reply:
Yeah